Introduce PSA test driver library to test PSA configuration

ronald-cron-arm opened the pull request:
## Description
The main purpose of this PR is to remove from the library (psa_crypto_*.c files) all the test code introduced for testing the PSA
configuration. This clean-up was highly desired for the 2.x LTS branch. This PR removes the aforementioned test code by replacing it by calls to a PSA test driver library. 

PR #3959 has proven that by adding a prefix (LIBTESTDRIVER1/libtestdriver1_ in this commit) to all MBEDTLS/PSA_* and medtls/psa_* symbols of a copy of the Mbed TLS library, we can build a library that can be linked with the Mbed TLS library.

This commit leverages this to build a PSA test driver library based on the Mbed TLS library code. Its purpose is to test the PSA
configuration where cryptographic operations are disabled in the library and provided by a test driver library.
    
The cryptographic features supported by the test library are defined by:
. a minimal configuration (in the sense of config.h), see config_test_driver.h
. PSA_WANT_* and PSA_ACCEL_* defined macros.

The PSA_WANT_* macros have to be the same as the ones used to build the Mbed TLS library the test driver library is supposed to be linked to as the PSA_WANT_* macros are used in the definition of structures and macros that are shared by the PSA crypto core, Mbed TLS drivers and the driver test library.

The PSA_ACCEL_* macros are intended to define the cryptographic features that have to be removed from the Mbed TLS library and thus supported by the test library in test scenarios. The PSA_ACCEL_* macros to build the test library are thus mirrored from the ones to build the Mbed TLS library by extended the crypto_config.h: see crypto_config_test_driver_entension.h.

Fixes #3680.

## Status
IN DEVELOPMENT

## Requires Backporting
Yes, forward porting to development

ronald-cron-arm commented on library/psa_crypto.c:
```diff
@@ -3211,6 +3212,9 @@ psa_status_t psa_asymmetric_decrypt( mbedtls_svc_key_id_t key,
 
     *output_length = 0;
 
+    if( ! PSA_ALG_IS_ASYMMETRIC_ENCRYPTION( alg ) )
```
I am not completely sure because the line numbers are not those of the last commit but it seems that the previous one is in the `psa_asymmetric_encrypt` API thus that the check is performed once at the beginning of each API which seems ok.

ronald-cron-arm commented on library/ssl_tls.c:
```diff
@@ -7528,7 +7528,7 @@ int mbedtls_ssl_get_key_exchange_md_tls1_2( mbedtls_ssl_context *ssl,
     }
 
-    if( ( status = psa_hash_finish( &hash_operation, hash, MBEDTLS_MD_MAX_SIZE,
+    if( ( status = psa_hash_finish( &hash_operation, hash, PSA_HASH_MAX_SIZE,
```
This code is "MBEDTLS_USE_PSA_CRYPTO" specific thus PSA_HASH_MAX_SIZE seems the right choice.

mpg commented:
@ronald-cron-arm does this still need work or did you just forget to remove the label after pushing?

ronald-cron-arm commented:
> @ronald-cron-arm does this still need work or did you just forget to remove the label after pushing?

I think the only genuine CI failure is the PSA compliance where a NOT_SUPPORTED error code is expected and we return INVALID_PARAMETER. Thus I still need to do something about that, leaning towards fixing the compliance test suite for the time being. But I do not think this should prevent the reviews to carry on and I believe I have addressed all comments from @mprse.

mprse reviewed (dismissed):
Preview issues are solved.
No new issues found.
Approve.

ronald-cron-arm commented:
I rebased on top of development_2.x to fix a conflict in psa_crypto.c introduced by the merge of #5231.

ronald-cron-arm commented:
I changed the "psa: Add algorithm validity check" commit to the "psa: asymmetric_encrypt/decrypt: Improve error code consistency" one to fix the PSA compliance test issue we were getting before.

gabor-mezei-arm commented on library/psa_crypto_mac.c:
```diff
@@ -30,18 +30,7 @@
 #include <mbedtls/error.h>
 #include <string.h>
 
```
In my opinion there should be a naming convention for the local functions. All of them should use the `_internal` suffix or none of them.

gilles-peskine-arm replied:
Historically, the `xxx_internal` functions are functions for which `xxx` was part of the API, and we needed part of its job to be a separate function. There's no general rule for what `_internal` indicates (the split in functionality is different for different `xxx`) and it would be good to have a more informative naming convention. However, I think this is out of scope of the current pull request. There are a number of pending follow-ups for this PR so I'd prefer for it to be merged as soon as possible, and to do naming improvements in a follow-up.

gilles-peskine-arm commented:
I am merging this pull request ahead of its 3.0 forward-port so that it's available for follow-up tasks such as #5249 and #5251, and so that there's no risk that it slips until after the 2.28 LTS.
